blog/_posts/2022-01-17-scala-3.1.1-released.md

```diff
@@ -0,0 +1,71 @@
+
+  ```scala
+    def example[A, B](a: A)(using A <:< B) =
+      val b: B = a
+      ???
+  ```
+
+  To know more see [PR #13662](https://github.com/lampepfl/dotty/pull/13662) and [related issue](https://github.com/lampepfl/dotty/issues/12955).
+- Mirrors can be correctly summoned for hierarchical sum types compiled by Scala 3.0.x ([PR #14100](https://github.com/lampepfl/dotty/pull/14100)).
+- Dual Scala 2/3 macros can be now declared in non-experimental scopes ([PR #13795](https://github.com/lampepfl/dotty/pull/13795)).
+- Scaladoc can correctly handle symbolic class names, such as `/` or `\` ([PR #13827](https://github.com/lampepfl/dotty/pull/13827)).
+- Scaladoc documentation has a new, improved UI.
+
+For a comprehensive list of fixes, take a look at the full changelog on GitHub. // TODO: link
+
+## 3.1.2-RC1 and forward compatibility improvements
+
+Scala 3 has excellent backward compatibility guarantees between the minor versions. On the other hand, after the recent release of Scala 3.1, we can see that the libraries should be really cautious with updating the compiler version, as it is forcing the bump on every user of that library. We do not want library authors to be stuck on old versions of the compiler as that would mean that they are locked out of many bugfixes, or we would need to spend enormous effort on backporting every bugfix to all past versioning lines.
+
+We are sure we want to mitigate those difficulties. We already have implemented the first and most important part of the solution and have it ready for public testing. Beginning in Scala 3.1.2-RC1, that was just released, compiler is able to generate output that can be consumed by older versions. To enable that, you need to specify a minimum version of the language using the experimental `-Yscala-release` flag.
+
+So, for example, compiling your library using Scala 3.1.2-RC2 with `-Yscala-release:3.0` flag, you make it possible for people still using Scala 3.0 to depend on your library. This has one limitation, you won't be able to use any symbol added to the standard library after 3.0. They are marked in the code and in the documentation by the internal `@since` annotation that was also added in this version.
+
+Those restrictions are not final, and in the future, we may decide to allow using some functions from the newer versions of the standard library and to make the `@since` annotation public.
+
+One important note is that the `-Yscala-release` flag is experimental and not yet available in any stable release of the compiler. It means the build tools, such as SBT or Mill, are not yet aware of its existence. Of course, you can declare usage of the flag in your build file, and the compiler will respect it. However, there may still be some hiccups, like generated artifacts depending on the incorrect version of the stdlib. We will do our best to help fix those minor issues before the stable release of 3.1.1.
```

Comment:
So what's the upshot for library authors? I'm a library author; what do I do, what do you recommend, what is expected of me, what is the upshot, what do I actually need to know about this right now?

I'd suggest leading with answers to those questions. _Then_ follow it up with technical details and justifications if you feel they are necessary.

Reading this, the situation seems to be rather complex and is still changing rapidly. In that case, it might actually be better to omit the technical details from the release announcement; consider giving people the TL;DR on where things stand, with a link to a contributors thread on Discourse where people can read about and discuss details.

(I might be wrong about what the right way to handle this is, but if I am wrong, that speaks to a lack of clarity in the current text, which should be addressed regardless.)

Reply:
```suggestion
One important note is that the `-Yscala-release` flag is experimental and not yet available in any stable release of the compiler. It means the build tools, such as SBT or Mill, are not yet aware of its existence. Of course, you can declare usage of the flag in your build file, and the compiler will respect it. However, there may still be some hiccups, like generated artifacts depending on the incorrect version of the stdlib. We will do our best to help fix those minor issues before the stable release of 3.1.2.
```
?

Reply:
this is improved. it's a bit hard for me to judge because I already know too much about the situation — I'd suggest running it by a few actual library maintainer or two who haven't been following this stuff closely